app/src/main/java/com/babylon/wallet/android/domain/usecases/p2plink/EstablishP2PLinkConnectionUseCase.kt

```diff
@@ -70,18 +69,25 @@ class EstablishP2PLinkConnectionUseCase @Inject constructor(
                 override suspend fun completeLinking(connectionId: String): Result<Unit> {
                     val p2pLinks = p2PLinksRepository.getP2PLinks()
 
-                    val walletClientKeyBytes = p2pLinks.findBy(payload.publicKey.hex)?.walletPrivateKey
-                        ?.decodeHex() ?: PrivateKey.EddsaEd25519.newRandom().toByteArray()
+                    val walletClientKeyBytes = encryptedPreferencesManager.getP2PLinksWalletPrivateKey()
+                        ?.decodeHex()
+                        ?: run {
+                            val newPrivateKey = PrivateKey.EddsaEd25519.newRandom().toByteArray()
+                            encryptedPreferencesManager.saveP2PLinksWalletPrivateKey(newPrivateKey.toHexString())
+                            newPrivateKey
+                        }
                     val walletClientKey = PrivateKey.EddsaEd25519.newFromPrivateKeyBytes(walletClientKeyBytes)
-
-                    val linkClientInteractionResponse = buildLinkClientInteractionResponse(
-                        walletClientPrivateKey = walletClientKey,
-                        connectionPassword = payload.password
+                    val linkClientInteraction = PeerdroidLink.LinkClientExchangeInteraction(
+                        publicKey = walletClientKey.publicKey(),
+                        signature = walletClientKey.signToSignature(
+                            hashedData = getSignatureMessageFromConnectionPassword(payload.password.value.hex)
+                        )
```

Comment:
Both of these are also existing in `NotaryPrivateKey` as methods. So maybe we need to introduce such class in Sargon.